projects/whydense/cifar-100/cifar_experiment.py

```diff
@@ -0,0 +1,529 @@
+    first_epoch_batch_size = config.get("first_epoch_batch_size", batch_size)
+    self.batches_in_epoch = config.get("batches_in_epoch", sys.maxsize)
+    self.batches_in_first_epoch = config.get("batches_in_first_epoch",
+                                             self.batches_in_epoch)
+
+    self.test_batch_size = config["test_batch_size"]
+    self.test_batches_in_epoch = config.get("test_batches_in_epoch", sys.maxsize)
+    self.noise_values = config.get("noise_values", [0.0, 0.1])
+    self.loss_function = nn.functional.cross_entropy
+
+    self.learning_rate = config["learning_rate"]
+    self.momentum = config.get("momentum", 0.5)
+    self.weight_decay = config.get("weight_decay", 0.0005)
+    self.learning_rate_gamma = config.get("learning_rate_gamma", 0.9)
+    self.last_noise_results = None
+    self.lr_step_schedule = config.get("lr_step_schedule", None)
+
+    # Network parameters
+    network_type = config.get("network_type", "vgg")
+    inChannels, self.h, self.w = config["input_shape"]
+
+    self.boost_strength = config["boost_strength"]
+    self.boost_strength_factor = config["boost_strength_factor"]
+    self.k_inference_factor = config["k_inference_factor"]
+
+    # CNN parameters - these are lists, one for each CNN layer
+    self.cnn_percent_on = config["cnn_percent_on"]
+    self.cnn_kernel_sizes = config.get("cnn_kernel_size", [3] * len(self.cnn_percent_on))
+    self.cnn_out_channels = config.get("cnn_out_channels", [32] * len(self.cnn_percent_on))
+    self.cnn_weight_sparsity = config.get("cnn_weight_sparsity",
+                                          [1.0] * len(self.cnn_percent_on))
+    self.in_channels = [inChannels] + self.cnn_out_channels
+    self.block_sizes = config.get("block_sizes", [1] * len(self.cnn_percent_on))
+    self.use_max_pooling = config.get("use_max_pooling", False)
+
+    # Linear parameters
+    self.linear_weight_sparsity = config["weight_sparsity"]
+    self.linear_n = config["linear_n"]
+    self.linear_percent_on = config["linear_percent_on"]
+    if isinstance(self.linear_n, int):
+      self.linear_n = [self.linear_n]
+      self.linear_percent_on = [self.linear_percent_on]
+      self.linear_weight_sparsity = [self.linear_weight_sparsity]
+    self.output_size = config.get("output_size", 10)
+    self.optimizer_alg = config.get("optimizer", 'SGD')
+
+    # Setup devices, model, and dataloaders
+    print("setup: Torch device count=", torch.cuda.device_count())
+    torch.manual_seed(seed)
+    if torch.cuda.is_available():
+      print("setup: Using cuda")
+      self.device = torch.device("cuda")
+      torch.cuda.manual_seed(seed)
+    else:
+      print("setup: Using cpu")
+      self.device = torch.device("cpu")
+
+    self.transform_train = transforms.Compose([
+      transforms.RandomCrop(32, padding=4),
+      transforms.RandomHorizontalFlip(),
+      transforms.ToTensor(),
+      transforms.Normalize((0.4914, 0.4822, 0.4465), (0.2023, 0.1994, 0.2010)),
```

Comment:
Just realized - isn't it incorrect to normalize after cropping? Shouldn't people normalize before cropping?